proxy/src/main/java/net/md_5/bungee/connection/DownstreamBridge.java

```diff
@@ -107,8 +107,9 @@ public void handle(PacketWrapper packet) throws Exception
     @Override
     public void handle(KeepAlive alive) throws Exception
     {
-        if (server.getSentKeepAlives().getFirst().getMillis() + bungee.getConfig().getTimeout() < System.currentTimeMillis()) {
-            throw new IllegalStateException("Did not recieve keepalive in time from " + con.getName());
+        if ( server.getSentKeepAlives().getFirst().getMillis() + bungee.getConfig().getTimeout() < System.currentTimeMillis() )
+        {
+            throw new KeepaliveTimeoutException( "Did not recieve keepalive in time from " + con );
```

Comment:
Just got the idea of simply modifying the exception handler of the downstream bridge to create identical behaviour as if the exception would've occurred in UpstreamBridge. @md-5